### ISSUE TITLE
Traceback AttributeError: id

### ISSUE THREAD

**Issue by masterlexa** [toluaina/pg-sync#30]
PGSync version: 1.1.7

Postgres version: 9.8

Elasticsearch version: 7.7.1

Redis version: latest

Python version: 3.7

Problem Description: 
I have traceback with sync

After I see a queue in Redis, which is full. Sync works but doesn't load rows to elastic.

Error Message (if any): 

```
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: 2020-08-17 21:18:50.522:ERROR:pgsync.sync: Exception id
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: Traceback (most recent call last):
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 210, in __getattr__
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     return self._data[key]
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: KeyError: 'id'
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: 
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: During handling of the above exception, another exception occurred:
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: 
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: Traceback (most recent call last):
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/sync.py", line 572, in pgsync.sync.Sync._sync
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/query_builder.py", line 804, in pgsync.query_builder.QueryBuilder.build_queries
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/query_builder.py", line 258, in pgsync.query_builder.QueryBuilder._children
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 212, in __getattr__
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     raise AttributeError(key)
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: AttributeError: id
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: 2020-08-17 21:18:50.620:ERROR:pgsync.sync: Exception: id
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: Traceback (most recent call last):
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 210, in __getattr__
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     return self._data[key]
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: KeyError: 'id'
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: 
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: During handling of the above exception, another exception occurred:
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: 
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: Traceback (most recent call last):
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/sync.py", line 644, in pgsync.sync.Sync.sync_payloads
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/elastichelper.py", line 59, in pgsync.elastichelper.ElasticHelper.bulk
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 425, in parallel_bulk
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     actions, chunk_size, max_chunk_bytes, client.transport.serializer
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 748, in next
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     raise value
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 121, in worker
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     result = (True, func(*args, **kwds))
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 140, in _helper_reraises_exception
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     raise ex
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 292, in _guarded_task_generation
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     for i, x in enumerate(iterable):
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 128, in _chunk_actions
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     for action, data in actions:
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/sync.py", line 575, in _sync
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/sync.py", line 572, in pgsync.sync.Sync._sync
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/query_builder.py", line 804, in pgsync.query_builder.QueryBuilder.build_queries
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "pgsync/query_builder.py", line 258, in pgsync.query_builder.QueryBuilder._children
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 212, in __getattr__
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]:     raise AttributeError(key)
Aug 18 00:18:50 elastics01 sync_product_shop/product_shop[19354]: AttributeError: id


```

**Comment by sivy** [comment]
I've seen this as well, though I had not raised the issue yet.

@masterlexa  can you provide some or all of your DB schema and pg-sync schema?

**Comment by toluaina** [comment]
I'm unable to reproduce this. Can you please share your DB and pgsync schema, please. 
@masterlexa

**Comment by sivy** [comment]
Here's a portion of my code:

SQLAlchemy table:

```python
surveys = Table(
    'surveys',
    metadata,
    Column('id', Integer, primary_key=True),
   # ...
}

sources = Table(
    'sources',
    metadata,
    Column('id', Integer, primary_key=True),
    Column('surveyid', Integer, ForeignKey('surveys.id')),
   # ...
)
```

pg-sync schema.json:

```json
[
    {
        "database": "mydb",
        "index": "surveys",
        "nodes": [
            {
                "table": "surveys",
                "schema": "public",
                "columns": [
                    "id"
                ],
                "children": [
                    {
                        "table": "sources",
                        "schema": "public",
                        "columns": [
                            "id",
                            "surveyid"

                        ],
                        "relationship": {
                            "variant": "object",
                            "type": "one_to_many"
                        }
                    }
                ]
```

**Comment by sivy** [comment]
SQL query (trimmed for columns):

```
pgsync_1              | SELECT JSON_BUILD_ARRAY(anon_1._keys) AS "JSON_BUILD_ARRAY_1",
pgsync_1              |        JSON_BUILD_OBJECT('id', surveys_1.id, 'sources', anon_1.sources) AS "JSON_BUILD_OBJECT_1",
pgsync_1              |        surveys_1.id
pgsync_1              | FROM public.surveys AS surveys_1
pgsync_1              | LEFT OUTER JOIN
pgsync_1              |   (SELECT CAST(JSON_BUILD_OBJECT('sources', JSON_AGG(JSON_BUILD_OBJECT('id', JSON_BUILD_ARRAY(sources_1.id)))) AS JSONB) AS _keys,
pgsync_1              |           JSON_AGG(JSON_BUILD_OBJECT('id', sources_1.id, 'surveyid', sources_1.surveyid)) AS sources,
pgsync_1              |           sources_1.surveyid AS surveyid
pgsync_1              |    FROM public.sources AS sources_1
pgsync_1              |    GROUP BY sources_1.surveyid) AS anon_1 ON anon_1.surveyid = surveys_1.id
pgsync_1              | WHERE surveys_1.id = 202
```

**Comment by sivy** [comment]
Exception traceback:

```
pgsync_1              | 2020-08-19 00:11:26.308:ERROR:pgsync.sync: Exception id
pgsync_1              | Traceback (most recent call last):
pgsync_1              |   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 210, in __getattr__
pgsync_1              |     return self._data[key]
pgsync_1              | KeyError: 'id'
pgsync_1              |
pgsync_1              | During handling of the above exception, another exception occurred:
pgsync_1              |
pgsync_1              | Traceback (most recent call last):
pgsync_1              |   File "pgsync/sync.py", line 572, in pgsync.sync.Sync._sync
pgsync_1              |   File "pgsync/query_builder.py", line 746, in pgsync.query_builder.QueryBuilder.build_queries
pgsync_1              |   File "pgsync/query_builder.py", line 252, in pgsync.query_builder.QueryBuilder._children
pgsync_1              |   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 212, in __getattr__
pgsync_1              |     raise AttributeError(key)
pgsync_1              | AttributeError: id
pgsync_1              | 2020-08-19 00:11:26.407:ERROR:pgsync.sync: Exception: id
pgsync_1              | Traceback (most recent call last):
pgsync_1              |   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 210, in __getattr__
pgsync_1              |     return self._data[key]
pgsync_1              | KeyError: 'id'
pgsync_1              |
pgsync_1              | During handling of the above exception, another exception occurred:
pgsync_1              |
pgsync_1              | Traceback (most recent call last):
pgsync_1              |   File "pgsync/sync.py", line 644, in pgsync.sync.Sync.sync_payloads
pgsync_1              |   File "pgsync/elastichelper.py", line 59, in pgsync.elastichelper.ElasticHelper.bulk
pgsync_1              |   File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 425, in parallel_bulk
pgsync_1              |     actions, chunk_size, max_chunk_bytes, client.transport.serializer
pgsync_1              |   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 748, in next
pgsync_1              |     raise value
pgsync_1              |   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 121, in worker
pgsync_1              |     result = (True, func(*args, **kwds))
pgsync_1              |   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 140, in _helper_reraises_exception
pgsync_1              |     raise ex
pgsync_1              |   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 292, in _guarded_task_generation
pgsync_1              |     for i, x in enumerate(iterable):
pgsync_1              |   File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 128, in _chunk_actions
pgsync_1              |     for action, data in actions:
pgsync_1              |   File "pgsync/sync.py", line 575, in _sync
pgsync_1              |   File "pgsync/sync.py", line 572, in pgsync.sync.Sync._sync
pgsync_1              |   File "pgsync/query_builder.py", line 746, in pgsync.query_builder.QueryBuilder.build_queries
pgsync_1              |   File "pgsync/query_builder.py", line 252, in pgsync.query_builder.QueryBuilder._children
pgsync_1              |   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 212, in __getattr__
pgsync_1              |     raise AttributeError(key)
pgsync_1              | AttributeError: id
```

**Comment by sivy** [comment]
Hope that's helpful @toluaina  @masterlexa

**Comment by masterlexa** [comment]
I cant share my prod DB here, sorry, but I see the same trouble @sivy.

**Comment by masterlexa** [comment]
I see for example with the scheme 
```
{
    "database": "my_db",
    "index": "company",
    "nodes": [
      {
        "table": "company",
        "schema": "public",
        "columns": [
          "company_name",
          "disabled",
          "image_logo",
          "image_preview",
          "inn",
          "url"
        ],
        "children": [
                    {
                        "table": "rating",
                        "schema": "public",
                        "columns": [
                          "total",
                          "comment_service",
                          "comment_actual_price",
                          "comment_actual_avail",
                          "comment_delivery",
                          "reliability"
                        ],
                        "relationship": {
                            "variant": "object",
                            "type": "one_to_one"
                        }
                    }
          ]
      }
    ]
  }

```

and 


```json
{
    "database": "my_db",
    "index": "product_shop",
    "nodes": [
      {
        "table": "product_shop",
        "schema": "public",
        "columns": [
          "source_id",
          "company_id",
          "name",
        ],
                "children": [{
                        "table": "offer_source",
                        "schema": "public",
                        "columns": [],
                        "relationship": {
                            "variant": "object",
                            "type": "one_to_one"
                        },
                        "children": [
                          {
                            "table": "discount",
                            "schema": "public",
                            "columns": [
                              "rate",
                              "start_at",
                              "stop_at",
                              "is_active",
                              "applicable"
                            ],
                            "relationship": {
                                "variant": "scalar",
                                "type": "one_to_one"
                            }
                          }
                        ]
        }
        ]
      }
    ]
  }
```

**Comment by sivy** [comment]
A note that I am seeing that this exception causes the data to not be indexes in Elasticsearch; After removing the child relationship the data is being indexed properly.

**Comment by toluaina** [comment]
This should be resolved in pgsync 1.1.8. @sivy @masterlexa 
I intend to expedite making the code base available

**Comment by alameddinc** [comment]
I think this issue contuniue
